driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.AuthenticationMechanism;
+import com.mongodb.MongoCredential;
+import com.mongodb.MongoInternalException;
+import com.mongodb.ServerAddress;
+import com.mongodb.lang.NonNull;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonBinary;
+import org.bson.BsonBinaryWriter;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonString;
+import org.bson.ByteBuf;
+import org.bson.RawBsonDocument;
+import org.bson.codecs.BsonDocumentCodec;
+import org.bson.codecs.EncoderContext;
+import org.bson.io.BasicOutputBuffer;
+
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslException;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.security.SecureRandom;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+
+import static java.lang.String.format;
+
+public class IamAuthenticator extends SaslAuthenticator {
+    private static final int RANDOM_LENGTH = 32;
+
+    public IamAuthenticator(final MongoCredentialWithCache credential) {
+        super(credential);
+    }
+
+    @Override
+    public String getMechanismName() {
+        AuthenticationMechanism authMechanism = getMongoCredential().getAuthenticationMechanism();
+        if (authMechanism == null) {
+            throw new IllegalArgumentException("Authentication mechanism cannot be null");
+        }
+        return authMechanism.getMechanismName();
+    }
+
+    @Override
+    protected SaslClient createSaslClient(final ServerAddress serverAddress) {
+        return new IamSaslClient(getMongoCredential());
+    }
+
+    private static class IamSaslClient implements SaslClient {
+        private final MongoCredential credential;
+        private int step = -1;
+        private byte[] clientNonce = new byte[RANDOM_LENGTH];
+        private String httpResponse;
+
+        IamSaslClient(final MongoCredential credential) {
+            this.credential = credential;
+        }
+
+        @Override
+        public String getMechanismName() {
+            AuthenticationMechanism authMechanism = credential.getAuthenticationMechanism();
+            if (authMechanism == null) {
```

Comment:
This check is redundant to the one that I asked to be added in the constructor for the outer class, so it can be removed.

Reply:
Done